File: plugin-sdk-ts/src/event.ts
import { callHost, callHostJson } from './host';
import type { PluginEventEnvelope } from './types';

/**
 * Emit an event that can be observed by the host and other plugins.
 */
export function emit(name: string, payload: unknown): void {
  callHost(
    'cognia_event_emit',
    JSON.stringify({ name, payload }),
  );
}

/**
 * Emit an event with a simple string payload.
 */
export function emitStr(name: string, message: string): void {
  emit(name, message);
}

/**
 * Get the current plugin's own ID.
 */
export function getPluginId(): string {
  const result = callHostJson<{ pluginId: string }>(
    'cognia_get_plugin_id',
    '',
  );
  return result.pluginId;
}

/**
 * Parse the host event-listener callback envelope received by `cognia_on_event`.
 */
export function parseEnvelope<TPayload = unknown>(
  input: string,
): PluginEventEnvelope<TPayload> | null {
  try {
    return JSON.parse(input) as PluginEventEnvelope<TPayload>;
  } catch {
    return null;
  }
}

File: plugin-sdk/src/event.rs
use crate::host;
use extism_pdk::*;

/// Emit an event that can be observed by the host and other plugins.
pub fn emit(name: &str, payload: &serde_json::Value) -> Result<(), Error> {
    let input = serde_json::json!({ "name": name, "payload": payload }).to_string();
    unsafe { host::cognia_event_emit(input)?; }
    Ok(())
}

/// Emit an event with a simple string payload.
pub fn emit_str(name: &str, message: &str) -> Result<(), Error> {
    emit(name, &serde_json::json!(message))
}

/// Get the current plugin's own ID.
pub fn get_plugin_id() -> Result<String, Error> {
    let result = unsafe { host::cognia_get_plugin_id(String::new())? };
    let parsed: serde_json::Value = serde_json::from_str(&result)?;
    Ok(parsed["pluginId"].as_str().unwrap_or("").to_string())
}
